drivers: timer: nrf: refactor for speed and correctness

pabigot opened the pull request:
The existing implementation of `nrf_rtc_timer:z_clock_set_timeout()` calculates the compare value based on a complex series of operations including an unconditional integer division and multiplication intended to ensure the compare value is aligned to a tick boundary.  On the nRF51 this division requires a call to an outline function with a data-dependent execution time.

In the common case where the timeout is configured less than one tick past the last observed tick the division can be elided, as can several extra operations intended to deal with fractional ticks.

The code also failed to correctly account for a ticks-per-cycle that violated the minimum delay required to guarantee a compare value would produce a compare event.  The minimum delay was also unreasonably long (about 1 ms).  Reduce it to a more reasonable value to allow for a higher ticks-per-second, and diagnose attempts to set the tick frequency above the supported maximum (8192 Hz).

Finally, move the parts of the compare calculation that are not dependent on the live counter value out of the locked region.

Prior to this change (but after a couple minor enhancements also present in this PR) the observed time between the `irq_lock()` and `irq_unlock()` in `z_clock_set_timeout()` on the nRF51 ranged between 5 us and 8 us.

With the revised algorithm the observed lock duration is between 2.16 us (1024 Hz) and 2.88 us (100 Hz) in the common case that the compare is set within the current tick.  If the compare is set late the duration will be higher, but no greater than the previous implementation.

These changes also remove significant instability with larger ticks-per-second (#12542) when `CONFIG_TIMESLICING=y` is present, allowing `CONFIG_TICKS_PER_SEC` to be increased to its maximum of 8192.

Closes #12541

pabigot committed:
drivers: timer: nrf: use irq_lock instead of spinlock

spinlock gains us nothing on an architecture that doesn't support SMP.
Use the standard irq_lock() API so when we search for conditions that
may decrease ISR responsiveness we can find them.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: timer: nrf: drop unnecessary counter mask

The RTC COUNTER register doesn't care that it receives a value larger
than it can hold; it'll discard the bits internally.  No need to spend
cycles doing it manually.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: timer: nrf: refactor for speed and correctness

The existing implementation of z_clock_set_timeout() calculates the
compare value based on a complex series of operations including an
unconditional integer division and multiplication intended to ensure the
compare value is aligned to a tick boundary.  On the nRF51 this division
requires a call to an outline function with a data-dependent execution
time.

In the common case where the timeout is set less than one tick past the
last observed tick the devision can be elided, as can several extra
operations intended to deal with fractional ticks.

The code also failed to account for a ticks-per-cycle that violated the
minimum delay required to guarantee a compare value would result in a
match without wrapping.  The minimum delay was also unreasonably long
(about 1 ms).  Reduce it to a more reasonable value to allow for a
higher ticks-per-second, and diagnose attempts to set the tick frequency
above the supported maximum (8192 Hz).

Finally, move the parts of the compare calculation that are not
dependent on the live counter value out of the locked region.

Prior to this change the observed time between the irq_lock() and
irq_unlock() in z_clock_set_timeout() on the nRF51 ranged between 5 us
and 8 us.

With the revised algorithm the observed lock duration is between 2.16 us
(1024 Hz) and 2.88 us (100 Hz) in the common case that the compare is
set within the current tick.  If the compare is set late the duration
will be higher, but no greater than the previous implementation.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -15,14 +15,26 @@
 
-static struct k_spinlock lock;
+#define COUNTER_MAX 0x00ffffffU
+#define MAX_TICKS (COUNTER_MAX / CYC_PER_TICK)
```
should be `((COUNTER_MAX - MIN_DELAY) / CYC_PER_TICK)`; update pending.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -15,14 +16,26 @@
+ * these calculations, allowing for a final check that can replace a
+ * too-low compare with a value that will guarantee fire.
+ */
+#define MIN_DELAY 4
```
The RTC IRQ could be pre-empted by Bluetooth and zero-latency IRQs.
IMHO we should an algorithmic solution here.

pabigot replied:
No, it can't.  Both the original spin_lock and the irq_lock() block the Bluetooth IRQs, which is why this is a problem.

pizi-nordic replied:
Ok. I missed the irq_lock().

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -33,7 +46,7 @@ static u32_t counter_sub(u32_t a, u32_t b)
 
 static void set_comparator(u32_t cyc)
```
Please use `static inline` or replace by direct call to nrf_rtc_cc_set().

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,21 +130,53 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+	 * cycles elapsed since the last tick.  Cap at the maximum
+	 * tick-aligned delta.
+	 */
+	u32_t cyc = min((1 + ticks) * CYC_PER_TICK, MAX_DELAY);
```
Doesn't it delay all announcements?
Is that workaround for rounding down in ms to tick conversion?

pabigot replied:
No.  It's a workaround for having to adjust for a non-zero time-since-last-tick in the locked region.  The only time it adds an "unnecessary" tick is if you get to this point in the code less than 30.6 us after RTC counter produced a compare event and the timer ISR completed update of `last_count`.  That's not going to happen very often.

pizi-nordic replied:
OK

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,21 +130,53 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+			cyc = MAX_DELAY;
+		}
 	}
+	set_comparator(last_count + cyc);
```
We should fire on tick boundary. Could we just keep adding CYC_PER_TICK to last_count until we met MIN_DELAY?

pabigot replied:
We do fire on a tick boundary.  `last_count` is tick-aligned, and `cyc` is a multiple of clocks-per-tick.

pizi-nordic replied:
I have no doubt about that. I am just curious, if we could replace a series of if's above by the simpler code.

pabigot replied:
I really don't think so.  The maximum time required for this code can be measured.  If we put a loop in, we have to worry about it iterating so many times that counter will have incremented more than twice and `MIN_DELAY` is no longer sufficient.

There are multiple conditions:
* We can be executing this within the tick recorded in `last_count` or in a later tick.  The first is the common case; the second requires expensive adds that are unrelated to `MIN_DELAY`, handled in the first inner condition.
* We can be more than `MIN_DELAY`  before the tick we're aiming for, or less.  The first is the common case; the second requires adds related to `MIN_DELAY`.  This can apply whether or not we needed to handle the first condition, and so is handled in the second condition.
* We can end up with cycles that exceed the counter representation.  This would break tick alignment unless ticks-per-sec evenly divides 32768.  We only have to worry about that if we did any adds, but in that case we can use a pre-computed maximum.

I don't think there's any way to simplify that.

pizi-nordic replied:
Ok.

pabigot replied:
We can, I think, change `>=` to `>` in the comparisons related to `MIN_DELAY`.  Lemme try that in the stress-test application.

pabigot replied:
I've adjusted the conditionals to be more strict, seems to work.  I'll let it keep running.

pabigot committed:
drivers: nrf: timer: add inline qualifier where inlining is intended

Not necessary with gcc, and Zephyr is inconsistent about using the
qualifier, but making the intent explicit is a good thing.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

andyross reviewed (approved):
Seems very reasonable.
